### ISSUE TITLE
Colour order (and therefore neume and syllable order) switched

### ISSUE THREAD

**Issue by JoyfulGen** [DDMAL/Neon#680]
The colour of two syllables were switched when I did the following steps (I've since tried repeating them and it happened again, so I'm pretty sure this sequence is in some way responsible for the switch):

1. I wanted to ungroup a syllable, but the ungrouping didn't work. This is because all the neume components of the syllable were part of the same neume; when this is the case Neon won't let you ungroup the syllable, because you can't have more than one syllable to a neume--totally normal so far.
2. I selected the neume and ungrouped it. The syllable highlight indicated that this was still one syllable (light blue), and the neume highlight showed four separate neumes, in the correct colour order. Still fine.
3. In this case, I wanted to group the neumes two-by-two, so I selected the first two and grouped them. I then selected the same two and grouped the neume components. At this point, the colour of the two neumes changed to green (it should've remained blue).
4. I then grouped the second two neumes and they remained blue. The order was now wrong, both of colours and syllables. The only way to fix it was to delete all four puncta and input them again.

When I tested the same thing on another folio, I tried grouping the neumes in the reverse order, starting with the last two. This removed the problem. So the issue only happens when I group the first group of neumes first. It looks like no matter what I do, Neon assigns the first neumes I group the "other" colour (in this case green). So if I group the last neumes first, there's no problem, but if I group the first neumes first, the assigned colour is wrong.

Here's what it looked like in folio 124r of Salzinnes:
<img width="917" alt="124r blue and green switched" src="https://user-images.githubusercontent.com/83373378/120658875-d46a5480-c453-11eb-9e3b-24a6ea801758.png">

You can see from the neumes before that light blue should be before green. Though I didn't screenshot the whole screen (oops), the syllable order in the text info section was also wrong.

The MEI: 
[124r blue and green switched.zip](https://github.com/DDMAL/Neon/files/6592219/124r.blue.and.green.switched.zip)

**Comment by JoyfulGen** [comment]
There was a more extreme example of switched colours in the same folio (124r), but it wasn't caused by the steps mentioned above. In fact I have no idea how it happened. So here it is:
<img width="1390" alt="124r fishy syllables illustrated" src="https://user-images.githubusercontent.com/83373378/120661204-0ed4f100-c456-11eb-95e0-d569ccc4f3b4.png">

The "et" circled in red is pink, and therefore is three colours off. I checked that all the syllables were assigned to the correct staves, and they were. The red arrow indicates where it should be, and the little red diamond is where it actually is. I fixed the problem by deleting all the neumes from the pink syllable to the light orange one and putting them back in again. So it's fixable, but it's still weird.

The MEI: 
[124r fishy colours (syllables).zip](https://github.com/DDMAL/Neon/files/6592390/124r.fishy.colours.syllables.zip)

**Comment by JoyfulGen** [comment]
Update for the first comment: it looks like I was wrong about the only fix for this being the deletion of the offending neumes. It would appear that when the colours of the neumes/syllables are wrong, the bounding boxes are actually in the correct colour order. @carrieeex discovered that if you drag the right bounding box to the left of the other one, the colours of the neumes flip! Occasionally, the colours of the BBs flip as well. (I really thought I had experimented with the placement of the bounding boxes, but it was apparently insufficient). So the neumes/syllables are now in the correct order, but the bounding boxes are not. However, dragging the bounding boxes back to their correct places doesn't affect the colour anymore, so everything is well again.

Here's what it looks like (in folio 132r of Salzinnes):
The initial situation, with the syllables in the wrong order and the BBs in the correct order:
<img width="254" alt="132r syllables no BBs yes" src="https://user-images.githubusercontent.com/83373378/120818695-d77f4680-c520-11eb-9b0d-ea458c7d8091.png">

After dragging the green BB to the left of the blue one, colours of syllables have flipped:
<img width="265" alt="132r syllables yes BBs no" src="https://user-images.githubusercontent.com/83373378/120818711-da7a3700-c520-11eb-9b40-b8ae63daffae.png">

After dragging the green BB back:
<img width="255" alt="132r all well" src="https://user-images.githubusercontent.com/83373378/120818714-dc43fa80-c520-11eb-8b4f-53f897e4c89a.png">

So I guess this isn't an issue anymore! But we still think it's weird and very counter-intuitive that the solution is to drag the BBs to the wrong place to get the syllables to flip to the correct order. It's also weird that moving BBs sometimes affects syllable order, and sometimes not.

**Comment by annamorphism** [comment]
The first part of this issue seems related to https://github.com/DDMAL/Neon/issues/657  
What is happening (as far as I can tell) is that the newly grouped neume components are always appended to the end of the syllable. The Bboxes seem to do something like the "grouping" of neumes does there, in that they "reset" the ordering to be as it is on the page. (I think taking the neume, moving it away, and putting it back has a similar effect.)

**Comment by annamorphism** [comment]
as for the problem on 124r another fix from the one @JoyfulGen did is to pull the neume or BBox out of place and back again. Or adding and deleting a neume.  I am not sure *why* it is appearing in this place though, unlike the other case.

**Comment by annamorphism** [comment]
This same bug is also true in browser (1c34e14)

**Comment by JoyfulGen** [comment]
> There was a more extreme example of switched colours in the same folio (124r), but it wasn't caused by the steps mentioned above. In fact I have no idea how it happened. So here it is:
> <img alt="124r fishy syllables illustrated" width="1390" src="https://user-images.githubusercontent.com/83373378/120661204-0ed4f100-c456-11eb-95e0-d569ccc4f3b4.png">
> 
> The "et" circled in red is pink, and therefore is three colours off. I checked that all the syllables were assigned to the correct staves, and they were. The red arrow indicates where it should be, and the little red diamond is where it actually is. I fixed the problem by deleting all the neumes from the pink syllable to the light orange one and putting them back in again. So it's fixable, but it's still weird.
> 
> The MEI:
> [124r fishy colours (syllables).zip](https://github.com/DDMAL/Neon/files/6592390/124r.fishy.colours.syllables.zip)

Another example of this happened in A02r (Salzinnes). Two syllables were off by two syllables (the order should be green-yellow-blue-pink-orange):
<img width="578" alt="A02r colour rave" src="https://user-images.githubusercontent.com/83373378/121913774-aca2a880-ccff-11eb-8c31-406caad3d13d.png">

The MEI file: [A02r colour order wonky.zip](https://github.com/DDMAL/Neon/files/6649336/A02r.colour.order.wonky.zip)

(There are other wonky things going on in this line, which I will post about after figuring out where to do so. TBC.)

**Comment by JoyfulGen** [comment]
The most extreme example of this phenomenon to date, in Salzinnes folio 131r, in Neon-staging. I numbered the syllables as they appear in the "text" section, to make the weird order apparent. There are also two syllables for which I can't find the bounding boxes and that therefore don't seem to exist (see https://github.com/DDMAL/MEI_encoding/issues/29), circled in red:
<img width="1391" alt="131r colour insanity (staging)" src="https://user-images.githubusercontent.com/83373378/124138178-631ec100-da54-11eb-8ec2-9c576803946b.png">
[131r syllable insanity (staging).zip](https://github.com/DDMAL/Neon/files/6748759/131r.syllable.insanity.staging.zip)
I tried processing the same folio in rodan2 and the exact same thing happened. I still don't know what causes this. Also, in this example, moving a bounding box doesn't change anything. The only fix appears to be to delete all the wonky syllables and input them again.

**Comment by annamorphism** [comment]
Another fix is splitting the staff, which then lets you move the syllables out and back again to produce the correct order (faster than inputting them all again. ) Not sure why this works though.

**Comment by JoyfulGen** [comment]
> Another fix is splitting the staff, which then lets you move the syllables out and back again to produce the correct order (faster than inputting them all again. ) Not sure why this works though.

That's great to know, thanks!

**Comment by carrieeex** [comment]
I had this issue (switched colour order) in A08r in Neon in Rodan2, which happens along with the [weird syllable issue](https://github.com/DDMAL/MEI_encoding/issues/29#issuecomment-875787723). 

I got this fixed with the following steps: 
after deleting the empty syllable in the MEI file, I uploaded the edited MEI in the browser Neon (I also tried in Rodan2, it is the same), and assign the wrong ordered syllable to another staff, then move it back, like this:
1. initially, "e" and "rat" in the wrong order
<img width="796" alt="A08v switched colour order -1 initially" src="https://user-images.githubusercontent.com/59303880/124804659-8bad2c00-df28-11eb-900d-c8af337f9bcd.png">
2. assign syllable "e" and "rat" to a different staff:
<img width="788" alt="A08v switched colour order -2 assign to a different staff" src="https://user-images.githubusercontent.com/59303880/124804665-8d76ef80-df28-11eb-8775-51d76220c729.png">
3. then put them back:
<img width="775" alt="A08v switched colour order -3 put it back" src="https://user-images.githubusercontent.com/59303880/124804671-8fd94980-df28-11eb-978a-de234ff86de1.png">
4. the order is now correct:
<img width="809" alt="A08v switched colour order -4 fixed" src="https://user-images.githubusercontent.com/59303880/124804680-91a30d00-df28-11eb-9d58-09965c5b84d6.png">

Something to notice is that it only works if [the weird syllable](https://github.com/DDMAL/MEI_encoding/issues/29#issuecomment-875787723) is deleted in the MEI file.

[A08v weird order and bbox - weird syllable deleted.zip](https://github.com/DDMAL/Neon/files/6779125/A08v.weird.order.and.bbox.-.weird.syllable.deleted.zip)
[A08v weird order and bbox.zip](https://github.com/DDMAL/Neon/files/6779126/A08v.weird.order.and.bbox.zip)

**Comment by annamorphism** [comment]
I have also had a case of colour order being incorrect at the staff level (see below, where there is no pink staff between the dark and light orange.) 

![image](https://user-images.githubusercontent.com/67451875/125508204-efe5f998-db76-47f2-831c-d710eafc613b.png)


(In the mei file at this spot there is ' sb n= "15" ' which isn't at any other staff break. Not sure if this is relevant)
![image](https://user-images.githubusercontent.com/67451875/125508487-52e8f83f-fb2b-49d4-9860-71afa7fb24ea.png)

A split At the end of the dark orange staff creates the missing pink staff: 
![image](https://user-images.githubusercontent.com/67451875/125510481-ba89f794-c377-4f7e-8588-b1314b9d969e.png)

But now moving the light orange away makes it blue, and moving the pink in turns it orange! 
![image](https://user-images.githubusercontent.com/67451875/125510577-6e346c24-a667-4fd7-b46b-d9bffe3b3416.png)

![image](https://user-images.githubusercontent.com/67451875/125510664-3eaf344c-9ac9-4652-926a-7c6d308c0675.png)


In other words there seems to be some sort of issue with the zones established on the facsimile which is interfering with the colour order.

**Comment by NicholasNorbertines** [comment]
I may be the only one still having issues related to this, so maybe I'm interpreting this wrong. However I noticed in a similar situation (I had tried to ungroup syllables, then ungrouped the neumes holding them together) that even when I entered new neume components, they were interpreted as falling in the wrong place. In other words, the neume *before* the inserted neumes was changing colour, but those after it were not. I didn't know to try moving the bboxes, but it was resolved when I deleted all of the offending neumes and reinserted them in sequence.

https://user-images.githubusercontent.com/99820491/191514512-423d5ce6-ffba-4b23-b264-c7dc00c32a83.mp4

**Comment by JoyfulGen** [comment]
@NicholasNorbertines I think in this situation the wrong colour order is most probably due to some of the neumes involved being grouped in the same syllable. If you insert a new neume between two neumes that are part of a single syllable, Neon has to understand that new neume as coming after that syllable, which is what makes the colours wonky. When this happens again, check the syllable groupings to see if this is the case. If so, ungroup whatever syllable(s) need to be ungrouped and the problem should be fixed. If the issue persists, please tell us!

**Comment by yinanazhou** [comment]
Hi @JoyfulGen, is this issue still happening?

**Comment by JoyfulGen** [comment]
Nope! Thanks for the heads-up, closing issue now.
